compiler/luci/pass/src/PropagateConcatenationQparamPass.cpp

```diff
@@ -0,0 +1,131 @@
+      node->quantparam()->scale.size() > 0 && node->quantparam()->zerop.size() > 0)
+    return true;
+
+  return false;
+}
+
+void overwrite_quantparam(luci::CircleConcatenation *concat, luci::CircleNode *target)
+{
+  auto concat_qparam = concat->quantparam();
+  assert(concat_qparam != nullptr);
+
+  auto target_qparam = target->quantparam();
+  assert(target_qparam != nullptr);
+  target_qparam->min = concat_qparam->min;
+  target_qparam->max = concat_qparam->max;
+  target_qparam->scale = concat_qparam->scale;
+  target_qparam->zerop = concat_qparam->zerop;
+  target_qparam->quantized_dimension = concat_qparam->quantized_dimension;
+}
+
+} // namespace
+
+namespace luci
+{
+
+void propagate_concat_quantparam(luci::CircleConcatenation *concat)
+{
+  // Check if concat is uint8-quantized
+  if (!is_uint8_quantized(concat))
+    return;
+
+  // Check if concat has no fused activation function
+  if (concat->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return;
+
+  const auto num_inputs = concat->numValues();
+  for (uint32_t i = 0; i < num_inputs; i++)
+  {
+    auto node = static_cast<luci::CircleNode *>(concat->arg(i));
+
+    // Skip if this input is not uint8-quantized
+    if (!is_uint8_quantized(node))
+      continue;
+
+    // Skip if concat and node have different quantized_dimension
+    if (node->quantparam()->quantized_dimension != concat->quantparam()->quantized_dimension)
+      continue;
+
+    // Skip if this input is CONCAT Op
+    if (node->opcode() == luci::CircleOpcode::CONCATENATION)
+      continue;
+
+    // Skip if this input is used by other Ops
+    auto succs = loco::succs(node);
+    if (succs.size() != 1)
+      continue;
+
+    assert(succs.find(concat) != succs.end());
+
+    overwrite_quantparam(concat, node);
+  }
+}
+
+/** BEFORE
+ *
+ *         [CircleNode]             [CircleNode]
+ *           (qparam1)                (qparam2)
+ *                   \                    /
+ *                    \                  /
+ *                    [CircleConcatenation]
+ *                          (qparam3)
+ *
+ *  AFTER
+ *         [CircleNode]             [CircleNode]
+ *           (qparam3)                (qparam3)
+ *                   \                    /
+ *                    \                  /
+ *                    [CircleConcatenation]
+ *                          (qparam3)
+ */
+bool PropagateConcatenationQparamPass::run(loco::Graph *g)
+{
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    auto concat = dynamic_cast<luci::CircleConcatenation *>(node);
+    if (not concat)
+      continue;
+
+    // Propagate qparam of concat to its inputs if
+    // (1) concat is uint8-quantized
+    // (2) concat has no fused activation function
+    // (3) the input is uint8-quantized
+    // (4) the input is not concatenation Op
+    // (5) the input is not produced to Ops other than concat
+    propagate_concat_quantparam(concat);
```

Comment:
It should also take into accout `CircleInput` and `CircleConst`.

Reply:
For `CircleInput`, I think it is ok to just overwrite qparam to the input node. The updated qparam is saved in the model, so users can give an input quantized with a proper qparam to the model.

For `CircleConst`,we have to quantize constant values according to propagated qparam. I'll add code for that. Thanks!

Reply:
> The updated qparam is saved in the model, so users can give an input quantized with a proper qparam to the model.

So, the user will have to requantize the input data because of the optimization? Well, I would not be happy ;D
Or I'm missing something?

Reply:
> So, the user will have to requantize the input data because of the optimization?

I think users do not have to 're'quantize the input data. Concat qparam is propagated when fp32 model is quantized to u8 model. Therefore users will only see the overwritten quantization parameters (users will not see the original qparam ever), which means input quantization is performed once.

Reply:
Oh, I see, thanks.